core/src/main/java/org/testcontainers/containers/GenericContainer.java

```diff
@@ -329,42 +331,40 @@ private void applyConfiguration(CreateContainerCmd createCommand) {
     }
 
     /**
-     * Wait until the container has started. The default implementation simply
-     * waits for a port to start listening; subclasses may override if more
-     * sophisticated behaviour is required.
+     * Specify the {@link WaitStrategy} to use to determine if the container is ready.
+     *
+     * @see Wait#defaultWaitStrategy()
+     * @param waitStrategy the WaitStrategy to use
+     * @return the {@code waitStrategy}, for chaining
      */
-    protected void waitUntilContainerStarted() {
-        waitForListeningPort(DockerClientFactory.instance().dockerHostIpAddress(), getLivenessCheckPort());
+    public WaitStrategy waitingFor(WaitStrategy waitStrategy) {
+        this.waitStrategy = waitStrategy;
+        return waitStrategy;
     }
 
     /**
-     * Waits for a port to start listening for incoming connections.
+     * The {@link WaitStrategy} to use to determine if the container is ready.
+     * Defaults to {@link Wait#defaultWaitStrategy()}.
      *
-     * @param ipAddress the IP address to attempt to connect to
-     * @param port      the port which will start accepting connections
+     * @return the {@link WaitStrategy} to use
      */
-    protected void waitForListeningPort(String ipAddress, Integer port) {
-
-        if (port == null) {
-            return;
-        }
-
-        try {
-            Unreliables.retryUntilSuccess((int) startupTimeout.getSeconds(), TimeUnit.SECONDS, () -> {
-                DOCKER_CLIENT_RATE_LIMITER.doWhenReady(() -> {
-                    try {
-                        new Socket(ipAddress, port).close();
-                    } catch (IOException e) {
-                        throw new RuntimeException(e);
-                    }
-                });
-                return true;
-            });
-        } catch (TimeoutException e) {
-            throw new ContainerLaunchException("Timed out waiting for container port to open (" + ipAddress + ":" + port + " should be listening)");
-        }
+    protected WaitStrategy getWaitStrategy() {
+        waitStrategy = ofNullable(waitStrategy).orElse(Wait.defaultWaitStrategy());
+        return waitStrategy;
     }
 
+    /**
+     * Wait until the container has started. The default implementation simply
+     * waits for a port to start listening; other implementations are available
+     * as implementations of {@link WaitStrategy}
+     *
+     * @see #waitingFor(WaitStrategy)
+     */
+    protected void waitUntilContainerStarted() {
+        getWaitStrategy()
+                .forContainer(this)
```

Comment:
`.forContainer()` should return new instance **or** should be removed in favour of an argument in `waitUntilReady` and stateless strategies.

Motivation - current implementation creates a false positive feeling that we can re-use existing strategy for different containers, but we can't.